### ISSUE TITLE
better documentation

### ISSUE THREAD

**Issue by ahicks92** [synthizer/synthizer#28]
We need:

- [ ] More tutorials
- [ ] A proper README
- [ ] Reorganize the manual to split up the object reference by category, and maybe fold some of the concepts in closer to the objects they're for.

And things with significantly less value that need to be decided on:

- [ ] do we want a blog?  I'm leaning toward no, but it isn't necessarily a bad idea.
- [ ] How much effort should be put into writing and maintaining a section of the manual for potential contributors? This worked well for the Rust compiler and there's enough subtlety that it might be useful for us, but it's unclear if that's enough to make up for needing to keep it up to date.

**Comment by chrisnorman7** [comment]
So I'd like to help, and this area seems like one I could do some good things in.

Do you have anywhere particular you'd like me to start? I'm thinking Tutorials, although they'd have to be in Python.

Is there is a particular folder you'd like stuff in, a particular style, that sort of thing?

**Comment by Tetsujinfr** [comment]
hi
I struggle to get started with HRTF. I had a quick look through the doc but I struggle to find my way.
I started with the python 3d media player, but after reading [this](https://synthizer.github.io/releases/0.8.x.html?highlight=hrtf#compatibility) I realised that stereo panning has been turned on by default recently, and I do not know how to enable HRTF back. The panning is quite brutal with the stereo mode, by default it is like balanced stereo if x=0, full left if x<0 and full right if x>0, with no smooth transition at all. Not sure if I am doing something wrong, but I am not sure what this example is meant to achieve without HRTF enabled by default anymore.

Then I tried to locate references to HRTF in the doc, there were not that many matches to my surpise. The closet page I found what the  [3D panning article](https://synthizer.github.io/concepts/3d_panning.html?highlight=hrtf#4-apply-the-panning-strategy), but I do not know how to use the HRTF panning mode still. I search for instance `SYZ_PANNER_STRATEGY_HRTF` in the git repo code base but could not find a match.

I had a quick look at the python examples in the example folder (fdn_reverb, echo and noise) but no evidence of HRTF settings in those neither.

Please help me, I am lost on how to enable and use HRTF with the lib. thanks a lot.

**Comment by Tetsujinfr** [comment]
Happy to help on writing a quick python tutorial complement, but I need your help to get started. I am not sure to understand if the SYZ_P_PANNER_STRATEGY is defined at the context level or the source level.

Let's say I start with the 3D media player python example below, how can setup the HRTF mode?
```
import sys
import synthizer
if len(sys.argv) != 2:
    print(f"Usage: {sys.argv[0]} <file>")
    sys.exit(1)

synthizer.configure_logging_backend(synthizer.LoggingBackend.STDERR)
synthizer.set_log_level(synthizer.LogLevel.DEBUG)
with synthizer.initialized():
    ctx = synthizer.Context( ???)
    ???

```
 
thanks a lot.

**Comment by ahicks92** [comment]
@chrisnorman7 
I missed your comment.  The answer is it's probably premature.  Let me finish breaking the library's API first.

@Tetsujinfr 
As I put in the other issue:

```
context.panner_strategy = synthizer.PannerStrategy.HRTF
```

And then create your sources.

**Comment by Tetsujinfr** [comment]
Thanks, all good now with enabling HRTF.
I got a few question regarding the HRTF rendering but I will open a new issue or leverage another one if it already exists.

**Comment by davidkreynolds** [comment]
Hi,
I'm currently attempting to write games for the visually impaired. Standardly, when writing a space shoot up, pitch dictates the Y cordinate, and pan the X coordinate. Can I have some clarification on the following:
1. I want to play a repeated sound (Very short), but depending on the location of the object making the sound, I need to adjust pitch and pan accordingly. The pitch and pan should relate to the relative position of the enemy and the player.
2. I did look at the manual, but could find nothing that my meager brain could interpret as a solution to this.
Many thanks,
David.

**Comment by ahicks92** [comment]
You need `BufferGenerator` with the `pitch_bend` property and `PannedSource` with the `panning_scalar` property.  Those two together should do what you want.

Avoid `StreamingGenerator` for this use case until #29 is closed.

**Comment by davidkreynolds** [comment]
Austin,

 

Thanks so much for this. At least I know what I’m looking for now,

 

David.

**Comment by davidkreynolds** [comment]
Hi Austin,
Can you point me in the direction of some usage/coding examples?
Thanks.

**Comment by ahicks92** [comment]
Clone the 0.9-dev branch and check examples or bindings/python/examples and you can see what this looks like once 0.9 lands.  That's all I've got.  @chrisnorman7 has at least one open source game written with it.

Can you articulate what's hard about this library?  The big example in the manual should be enough to figure most things out if combined with the object references.  There's not supposed to be anything complicated enough here to need a lot of detailed demonstration, at least not until custom streaming stuff in 0.9 that's not out yet.

If your issue is a lack of Python-specific documentation, this isn't going to be fixed: I don't have the bandwidth to maintain two sets of everything.  Shortly once I take over the Rust bindings that will be 3.  It doesn't scale for a weekend project.  But the C API translates in a very straightforward manner and you should be able to figure it out by reading the cython or even just playing with it in a shell.

**Comment by davidkreynolds** [comment]
Austin,

 

I completely get it. It’s a little tricky, as the frequency of the sound needs to be linked to the movement of an object, as does the pan position.

Further to this, can I have more than one sound playing at the same time?

I’ll take another look at the big example in the docs…

**Comment by ahicks92** [comment]
Yeah, of course you can have more than one sound.  You just make more sources, or hook more generators to one source, depending.  For what it's worth I am blind, and I am writing this to try to make an MMO for the blind, so I already know what you're trying to do.

pitch_bend is a multiplicative factor.  That is to say that if you make a buffer of a tone at 300 hz and you put it on a generator with pitch_bend=2.0, you get a tone at 600 hz.  If you're looking for an audio library that can take a sound and frequency shift it to a given frequency directly, you're not going to find one that can do so in realtime.  You have to work out the ratio yourself and have whoever is doing sound design make sounds with the same primary frequency so that you know what this should be.

For what it's worth, this issue isn't about answering random questions. It's about collecting documentation requirements and tracking documentation improvements.  Can I assume that you're either (1) new to audio libraries or (2) new to programming, or is Synthizer really that much more confusing than other options?  I am trying to determine how to turn our discussion into something actionable.

**Comment by davidkreynolds** [comment]
Hi Austin,

 

I’m new to sound libraries. Thanks for explaining the multipative technique.

I get where you’re coming from, and I’ll bother you no more on this issue,

 

Many thanks,

David.

**Comment by ahicks92** [comment]
I don't mind being bothered, and I probably need to set up a proper support mechanism.  I'm just trying to direct stuff that's not related to the issue tracker out of the issue tracker.  For now, if you need further help my e-mail is ahicks@ahicks.io.  And of course feel free to open issues if you find bugs.

I'll look at getting a google groups or something going in the next few days.  I really should have done it a long time ago.

**Comment by ahicks92** [comment]
Closing this out because I just rewrote the manual, massively expanding it in the process.  Taking it further than I now have is infeasible.
